### ISSUE TITLE
Grid is not switch back to normal view (flat data) after pivot is applied and then turned off

### ISSUE THREAD

**Issue by Krzysztof1985** [ag-grid/ag-grid#4940]
```
[x] bug report => see 'Providing a Reproducible Scenario'
[] feature request => do not use Github for feature requests, see 'Customers of AG Grid'
[x] support request => see 'Requesting Community Support'
```

**Requesting Community Support**
Hello I am using AG-grid and I am  using to use server **side row data model**. Based on AG-Grid-Oracle example I adjusted a little bit the queries and behaviour for postgres db, after some filter adjustments on backend in Java (my customization  in implementation) works nicely!
I have an issue with pivot mode. When I switch to pivot mode, perform some operations and then I switch off (using slider) the pivot mode - the grid is not refreshing the layout, and it's not turn back to normal state. (flat data)
I was trying to manage this using saving the default grid state and definition on startup (onFirstRendered)
and then with pivot mode event is fired, I am checking is switched off and then I tried to apply the default state, or something like this:
```javascript
      const columnOrder = gridColumnApi
        .getAllDisplayedColumns()
        .map(column => column.getColId());
      gridColumnApi.moveColumns(columnOrder, 0);
```
but the UI is still presenting me pivot mode with no data.

This is my code snippet. There is no issue with backend. I tried to enforce refresh the AG-Grid by gridApi, by gridRef.
Nothing is working.

**Providing a Reproducible Scenario**
Accepted reproducible scenarios are
- Open AG-Grid with Server side rod data model,
- click on slider to activate pivot
- perform some grouping (2-3 is fair enough)
- click label column
- click value column (pivot is visible and works OK)
- turn off the pivot mode using slider
![pivotMode-state](https://user-images.githubusercontent.com/2550886/155560341-5efb2c57-a461-4dfa-a122-e286edfdcf2e.png)

**Current behavior**

Grid is holding pivot view mode.

**Expected behavior**
Grid is switch back to default initial view/mode (flat data)
![pivot-mode-seq2](https://user-images.githubusercontent.com/2550886/155560397-fd958586-2b92-4230-89bd-e0e3e0fcdd2e.png)

```javascript
//required imports

const Worksheet = () => {
  const gridRef = useRef();
  
  const [defaultColumnState, setDefaultColumnState] = useState(null);
  const [defaultGroupState, setDefaultGroupState] = useState(null);
  const [defaultSortModel, setDefaultSortModel] = useState(null);
  const [defaultFilterModel, setDefaultFilterModel] = useState(null);
  const [defaultValueColumns, setDefaultValueColumns] = useState(null);
  const [enableRangeSelection] = useState(true);
  const [suppressAggFuncInHeader] = useState(true);
  const [sideBar] = useState(true);
  const [groupIncludeTotalFooter] = useState(true);
  const [groupIncludeFooter] = useState(true);
  const [showEnableRowGroupPanel] = useState("always");
  const [animateRows] = useState(true);
  const [rowModelType] = useState("serverSide");
  const [serverSideStoreType] = useState("partial");
  const [gridApi, setGridApi] = useState(null);
  const [gridColumnApi, setGridColumnApi] = useState(null);
  
  const handlePivotMode = () => {
    const isPivotMode = gridColumnApi.isPivotMode();
    if (isPivotMode === false) {

      // // gridRef.current.api.columnApi.setColumnGroupState(defaultGroupState);
      // gridColumnApi.setColumnGroupState(defaultGroupState);
      // gridColumnApi.setValueColumns(defaultValueColumns);
      // // gridRef.current.api.columnApi.setValueColumns(defaultValueColumns);
      //
      // gridRef.current.api.setFilterModel(defaultFilterModel);
      // gridRef.current.api.setSortModel(defaultSortModel);
      //
      // gridRef.current.api.columnApi.applyColumnState({
      //   state: [defaultColumnState],
      //   defaultState: {
      //     pivot: false,
      //     rowGroup: false,
      //   },
      // });

      const columnOrder = gridColumnApi
        .getAllDisplayedColumns()
        .map(column => column.getColId());
      gridColumnApi.moveColumns(columnOrder, 0);

      // gridApi.setFilterModel(defaultFilterModel);
      // gridApi.columnApi.applyColumnState(defaultSortModel);
      // gridColumnApi.setColumnGroupState(defaultGroupState);
      // gridColumnApi.setValueColumns(defaultValueColumns);
      // gridColumnApi.applyColumnState({
      //   state: [defaultColumnState],
      //   defaultState: {
      //     pivot: false,
      //     rowGroup: false,
      //   },
      // });
    }
  };
  
  const handleOnGridReady = params => {
    setGridApi(params.api);
    setGridColumnApi(params.columnApi);
    params.api.setServerSideDatasource(datasource);
  };

  const datasource = {
    getRows(params) {
      backendService.getRows(params.request)
        .then(response => {
          params.successCallback(response.data, response.lastRow);
          params.success({ rowData: response.data });
          addPivotColDefs(response, params.columnApi);
        })
        .catch(error => {
          console.log(error);
          params.failCallback();
        });
    },
  };

  const onFirstDataRendered = () => {
    setDefaultColumnState(gridColumnApi.getColumnState());
    setDefaultColumnState(gridColumnApi.getColumnState());
    setDefaultGroupState(gridColumnApi.getColumnGroupState());
    setDefaultValueColumns(gridColumnApi.getValueColumns());
    setDefaultSortModel(gridApi.getSortModel());
    setDefaultFilterModel(gridApi.getFilterModel());
  };

  const addPivotColDefs = (response, columnApi) => {
    const existingPivotColDefs = columnApi.getSecondaryColumns();
    if (existingPivotColDefs && existingPivotColDefs.length > 0) {
      return;
    }
    const pivotColDefs = response.secondaryColumnFields.map(function(field) {
      const headerName = field.split("_")[0];
      return { headerName, field };
    });
    columnApi.setSecondaryColumns(pivotColDefs);
  };

  return (
      <AgGridReact
        ref={gridRef}
        maxConcurrentDatasourceRequests={1}
        columnDefs={columnsDefinition}
        autoGroupColumnDef={autoGroupColumnDef}
        rowModelType={rowModelType}
        defaultColDef={defaultColumnDefinition}
        onGridReady={handleOnGridReady}
        serverSideStoreType={serverSideStoreType}
        sideBar={sideBar}
        enableRangeSelection={enableRangeSelection}
        cacheBlockSize={100}
        suppressAggFuncInHeader={suppressAggFuncInHeader}
        groupIncludeTotalFooter={groupIncludeTotalFooter}
        groupIncludeFooter={groupIncludeFooter}
        rowGroupPanelShow={showEnableRowGroupPanel}
        onColumnPivotModeChanged={handlePivotMode}
        animateRows={animateRows}
        pivotColumnGroupTotals="before"
        onFirstDataRendered={onFirstDataRendered}
        rowGroupOpened={handleRowGroupOpened}
      />
  );
};
```

**Please tell us about your environment:**
Mac OS 10, IntelliJ, Java 11 + Spring boot 2.5, PostgreSQL

* **AG Grid version:** 26.2.0

* **Browser:**
Chrome  98.0.4758.109 (x86_64)

* **Language:** [ReactJS]

Please help me how to solve this issue correctly :)

**Comment by icedjello** [comment]
Hi, 

Thank you for raising this. 

A member of our team will have a look and get back to you as soon as we have an answer. 

Thanks again

**Comment by icedjello** [comment]
Hi, 

I've tried to replicate this behaviour in the following plunker example: 

https://plnkr.co/edit/1BzwsWcUmstQjgsR  

I followed your steps: 

- Open AG-Grid with Server side rod data model,
- click on slider to activate pivot
- perform some grouping (2-3 is fair enough)
- click label column
- click value column (pivot is visible and works OK)
- turn off the pivot mode using slider
- pivotMode-state

And the grid is working as expected.

Are you able to replicate your issue in the example above?

**Comment by makinggoodsoftware** [comment]
Hi,

 

This ticket has been flagged as managed-by-community/waiting-for-repro for a while now and there has not been any activity, to help us tidy up we are going to close it.

 

Thanks

**Comment by Nestiii** [comment]
Hey @Krzysztof1985 ,  did u managed to solve it? Im having exactly the same issue.

**Comment by Adam-predictx** [comment]
> Hey [@Krzysztof1985](https://github.com/Krzysztof1985) , did u managed to solve it? Im having exactly the same issue.

Hi Nestiii, did you manage to fix this?

**Comment by Krzysztof1985** [comment]
> Hey [@Krzysztof1985](https://github.com/Krzysztof1985) , did u managed to solve it? Im having exactly the same issue.

Unfortunately no 😔

**Comment by Nestiii** [comment]
> > Hey [@Krzysztof1985](https://github.com/Krzysztof1985) , did u managed to solve it? Im having exactly the same issue.
> 
> Hi Nestiii, did you manage to fix this?

Nope, i did some ugly shenanigans to render it correctly

**Comment by Adam-predictx** [comment]
> > > Hey [@Krzysztof1985](https://github.com/Krzysztof1985) , did u managed to solve it? Im having exactly the same issue.
> > 
> > 
> > Hi Nestiii, did you manage to fix this?
> 
> Nope, i did some ugly shenanigans to render it correctly

Ah could you point me in that direction; been stuck on this for a few days

**Comment by isaiahshiner** [comment]
@Adam-predictx 
I wasn't able to fully reproduce the issue, but I made some progress and I'll share my notes.

TLDR: if `columnDefs` change in the same re-render as `pivotMode` changing to false, the columns disappear.

https://github.com/user-attachments/assets/3b8f4429-3023-4f7c-8859-a548ef7da380

Caveats:

1. We're still on version 32.0.0. 
   1. Yes, I know... But I haven't been able to repro on a Plunker on the same version either.
   2. We're updating soon™️, I'll note if it actually fixes our issue
   3. Update: I tried updating as far as 32.3.9, but it didn't help.
4. I don't experience the exact problem the OP shows.
   1. My grid still renders flat data correctly. 
   2. But I do see the columns in the sidebar disappear, specifically when turning `pivotMode` off.

The order of operations look like this:

1. The table starts with `pivotMode` true and correctly rendering pivoted data
2. User clicks the pivotMode toggle
3. `onColumnPivotModeChanged` is called and and the grid container handles the callback.
   1. The grid container is actually several jQuery components (yea, blah). So we also do a `ReactDOM.render` before we get back to AgGrid. No idea if that matters.
   2. This also saves a bunch of server side state, like pivotMode, then passes it back down.
4. The actual component that renders `AgGridReact` starts re-rendering
5. The `columnDefs` are calculated during rendering, and is basically a new array every render.
   1. There is a `useMemo`, but we created a bug that meant it wasn’t working correctly.
   2. This also occurred a different way, when we tried making the `pivot` attribute of each `columnDef` respect pivotMode, so they would be false if pivotMode was false. This seems sensible, but causes the same bug.
   3. In fact, I reproed this simply by assigning `columnDefs[0].nonsense = Math.random()`. It appears that as long as `!_.isEqual(prevColumnDefs, columnDefs)`, and pivotMode is changing to false, the bug will occur.
6. AgGridReact re-renders, and the columns in the sidepanel disappear.
   1. This does apparently happen before any other re-renders or side effects happen
7. Over several more re-renders and then calls to SSRM `getRows`, the table renders with correct data, but the columns are still gone.
8. Forcing the `columnDefs` to recalculate (in my case, by resizing a column width) makes the columns come back.

I'm probably going to hack fix my issue like this:

```typescript
let columnDefs = useColumnDefHook(tonsOfServerSideStateStuff);
if (prevPivotMode === true && pivotMode === false) {
    columnDefs = prevColumnDefs
}
```
I think a correctly implemented `useMemo` would probably also work, but we don't have a strong culture on respecting `react-hooks/exhaustive-deps` right now 🤦 so I worry the problem would just get added back later.
